tensordict/nn/storage.py

```diff
@@ -0,0 +1,421 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+import abc
+from abc import abstractmethod
+from typing import Callable, Dict, Generic, List, TypeVar
+
+import torch
+
+import torch.nn as nn
+
+from tensordict import NestedKey, TensorDict, TensorDictBase
+from tensordict.nn.common import TensorDictModuleBase
+
+K = TypeVar("K")
+V = TypeVar("V")
+
+
+class TensorStorage(abc.ABC, Generic[K, V]):
+    """An Abstraction for implementing different storage.
+
+    This class is for internal use, please use derived classes instead.
+    """
+
+    @abstractmethod
+    def clear(self) -> None:
+        raise NotImplementedError
+
+    @abstractmethod
+    def __getitem__(self, item: K) -> V:
+        raise NotImplementedError
+
+    @abstractmethod
+    def __setitem__(self, key: K, value: V) -> None:
+        raise NotImplementedError
+
+    @abstractmethod
+    def __len__(self) -> int:
+        raise NotImplementedError
+
+    @abstractmethod
+    def contain(self, item: K) -> torch.Tensor:
+        raise NotImplementedError
+
+
+class DynamicStorage(TensorStorage[torch.Tensor, torch.Tensor]):
```

Comment:
what about making this or the parent class a nn.Module?
I'm suggesting this because there is a tensor `default_tensor` in it, and it may be useful to register it as a buffer such that calling `dyn_storage.to("cuda")` sends the tensor to the device.